### PR TITLE
[Train] Add `PrintCallback`

### LINKED ISSUE #20806: [Train] Add `PrintingCallback`
Add a simple `PrintingCallback` to print out intermediate results from the training run.

Provides a cleaner approach to viewing the info passed into `train.report(...)` instead of having to add 

```
if train.world_rank() == 0:
    print(...)
```

in the training function.

### PR DESCRIPTION
**Pull request by bveeramani**
## Related issue number

Closes #20806 

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/train/callbacks/logging.py [hunk 1]
```diff
@@ -19,6 +20,42 @@
 logger = logging.getLogger(__name__)
 
 
+class PrintCallback(TrainingCallback):
```
Comment (matthewdeng) [inline python/ray/train/callbacks/logging.py:23]: Move this to a `print.py`?

File: python/ray/train/callbacks/logging.py [hunk 2]
```diff
@@ -19,6 +20,42 @@
...
+        """
+        # Check that the dictionaries have the same keys.
+        assert all(result.keys() == results[0].keys() for result in results)
+
+        dictionary = collections.defaultdict(list)
+        for result in results:
+            for key in result:
+                dictionary[key].append(result[key])
+
+        print(json.dumps(dictionary, indent=4))
```
Comment (matthewdeng) [inline python/ray/train/callbacks/logging.py:56]: Let's just keep this simple and print out the raw results for now (similar to the JSON logger callback).

We can add a note in the PR description/elsewhere to keep track of this as a possible enhancement.

Comment (bveeramani) [reply python/ray/train/callbacks/logging.py]: Addressed

File: python/ray/train/callbacks/logging.py (hunk 1 above)
Comment (bveeramani) [reply python/ray/train/callbacks/logging.py]: Addressed

**Comment by bveeramani** [comment]
The lint status check is failing because there's apparently a formatting issue with `logging.py`. I ran `./scripts/format.sh` but it didn't reformat `logging.py`. I also manually checked the yapf version and ran `yapf --style .style.yapf ./python/ray/train/callbacks/logging.py`, but it also didn't make any changes. Maybe some non-deterministic yapf nonsense?

Will debug when I get back from OOO.

**Review by matthewdeng — APPROVED** [review]
LGTM @amogkam can you merge?

File: python/ray/train/callbacks/print.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
+import json
+from typing import Dict, List
+
+from ray.train.callbacks import TrainingCallback
+
+
+class PrintCallback(TrainingCallback):
+    """A callback that prints training results to STDOUT.
+
+    Example:
```
Comment (amogkam) [inline python/ray/train/callbacks/print.py:10]: Can we change the example to show usage with `Trainer`? Users shouldn't be calling `handle_result` directly, so I think this docstring could be confusing.

**Review by amogkam — CHANGES_REQUESTED** [review]

File: python/ray/train/callbacks/print.py (hunk 3 above)
Comment (matthewdeng) [reply python/ray/train/callbacks/print.py:10]: Good point, we can also take the `handle_result` example and move it to the `handle_result` docstring?

File: python/ray/train/callbacks/print.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        >>> from ray import train
+        >>> from ray.train import Trainer
+        >>> from ray.train.callbacks import PrintCallback
+
+        >>> def train_func():
+        ...    for i in range(2):
+        ...        train.report(worker_idx=train.world_rank(), epoch=i)
+
+        >>> trainer = Trainer(num_workers=2, backend="torch")
+        >>> trainer.run(train_func, callbacks=[PrintCallback()])
```
Comment (matthewdeng) [inline python/ray/train/callbacks/print.py:21]:
```suggestion
        >>> trainer = Trainer(num_workers=2, backend="torch")
        >>> trainer.start()
        >>> trainer.run(train_func, callbacks=[PrintCallback()])
```

File: python/ray/train/callbacks/print.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+    Example:
+
+        >>> from ray import train
+        >>> from ray.train import Trainer
+        >>> from ray.train.callbacks import PrintCallback
+
+        >>> def train_func():
+        ...    for i in range(2):
+        ...        train.report(worker_idx=train.world_rank(), epoch=i)
```
Comment (matthewdeng) [inline python/ray/train/callbacks/print.py:18]: nit: Preference on `>>>` vs. `..code-block` formatting? Personally prefer to just be able to copy/paste the example and run it.

File: doc/source/train/user_guide.rst [hunk 6]
```diff
@@ -455,7 +456,7 @@ the callback is general enough, please feel welcome to `add it <https://docs
 .ray.io/en/master/getting-involved.html>`_ to the ``ray``
 `repository <https://github.com/ray-project/ray>`_.
 
-A simple example for creating a callback that will print out results:
+A simple example for creating a callback that will print out the average of all results:
```
Comment (matthewdeng) [inline doc/source/train/user_guide.rst:459]: Example should be changed accordingly?

File: python/ray/train/callbacks/print.py (hunk 4 above)
Comment (amogkam) [reply python/ray/train/callbacks/print.py]: Fixed!

File: python/ray/train/callbacks/print.py (hunk 5 above)
Comment (amogkam) [reply python/ray/train/callbacks/print.py:18]: My reasoning was that the `>>>` formatting would be easier to show what's actually being printed out. If we do `..code-block` then we would have to show the stdout as a comment. But good point on being able to copy paste. I don't really have a strong preference here, so am happy to change it to code block.

File: doc/source/train/user_guide.rst (hunk 6 above)
Comment (amogkam) [reply doc/source/train/user_guide.rst:459]: reverted this back

**Review by amogkam — APPROVED** [review]

**Pull request merged**